src/inmanta/agent/cache.py

```diff
@@ -63,51 +62,33 @@ def __init__(
             a callable expecting the cached value as an argument.
         """
         self.key = key
-        self.scope = scope
+        self.scope: Scope = scope
         self.value = value
-        self.time: float = time.time() + scope.timeout
         self.call_on_delete = call_on_delete
+        self.expiry_time = time.time() + scope.timeout
+        self.called_finalizer = False
```

Comment:
I added `self.called_finalizer` with the intention to make sure we don't call it twice because the following test was failing without it, but I very well might be missing something:

https://github.com/inmanta/inmanta-core/blob/b6146e6f79c061fb533fdeb79d9621c89740893d/tests/test_cache.py#L244

```
>       assert alpha.deleted == alpha.created
E       assert 2 == 1
E        +  where 2 = <test_cache.test_multi_threaded.<locals>.Spy object at 0x76113933c4d0>.deleted
E        +  and   1 = <test_cache.test_multi_threaded.<locals>.Spy object at 0x76113933c4d0>.created
```

Reply:
As far as I can tell the current implementation would indeed fix it in most cases, but since you have a non-thread safe check and write to this attribute, I think a race condition is possible when two threads call the method at roughly the same time.

That is, unless we are confident that that would never happen. If it's really about garbage collection and *one* explicit call, we can perhaps assume that garbage collection will only take place after we've finished all our calls, because we would still have a reference to the object while calling delete. But if we rely on that I would very clearly state why the variable is there and why it's safe without thread-safe access.

Reply:
I added a lock to make read/write to `called_finalizer` thread-safe